### PR TITLE
Hybrid class-instances and instance-classes

### PR DESCRIPTION
**Pull request by VitWW**
This proposal adds hybrid class-instances and instance-classes

It is an alternative to #461

[Rendered](https://github.com/VitWW/ghc-proposals/blob/hybrid-class-instances/proposals/0000-hybrid-class-instances.rst)

### CODE REVIEW

**Review by konsumlamm — COMMENTED** [review]
If I understand correctly, the essence of this proposal is that you want a way to define an instance without having to define instances for all the superclasses separately? I think `class instance` and `instance class` are really bad names, they don't mean anything to me.

File: proposals/0000-hybrid-class-instances.rst [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+.. author:: Viktor WW
+.. date-accepted::
+.. ticket-url:: 
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/593>`_.
+.. sectnum::
+.. contents::
+
+This proposal helps refactoring classes and instances such that axioms / rules and emergent properties are separated, but they are easy combined
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-hybrid-class-instances.rst:13]: For example: [Algebraic structures](https://upload.wikimedia.org/wikipedia/commons/0/07/Algebraic_structures_-_magma_to_group.svg)
Lines are "given properties", names are "emergent properties".
From 3 given properties we could construct 9 emergent properties.

Ok, I'll change this formulation.

File: proposals/0000-hybrid-class-instances.rst [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+
+Instance-Class Semantics
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+If we write ``instance class (C1 a, C2 a, ... Cn a) => IC a`` then:
+
+1) instance-class ``IC`` is a like an instance, so it must have no new methods
+2) Rules (2) ... (7) of instance-class ``IC`` has same properties as class-instance ``CI``
+3) instance-class ``IC`` are constraint synonym and we suggest to use it in this role
```
[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-hybrid-class-instances.rst:75]: Do I miss something? I could write `type IC a = (C1 a, C2 a, C3 a)`? Which extension allow that?

If it is true, then we could remove (3) rule

[the remark this answers is not in the document]
Comment (konsumlamm) [reply proposals/0000-hybrid-class-instances.rst:75]: Yes, you can write that with the [ConstraintKinds](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/constraint_kind.html#extension-ConstraintKinds) extension.

[the remark this answers is not in the document]
Comment (VitWW) [reply proposals/0000-hybrid-class-instances.rst:75]: Thanks
```haskell
type Stringy a = (Read a, Show a)
```

**Comment by VitWW** [comment]
So, I close this RFC for a revision

Thanks everyone for feedback!

**Pull request closed without merging**